engine/src/main/java/org/hibernate/search/backend/impl/DeletionQueryMapper.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.impl;
+
+import org.apache.lucene.search.Query;
+import org.hibernate.search.backend.DeletionQuery;
+import org.hibernate.search.util.impl.ScopedAnalyzer;
+
+/**
+ * @author Martin Braun
+ */
+public interface DeletionQueryMapper {
+
+	Query toLuceneQuery(DeletionQuery deletionQuery, ScopedAnalyzer analyzerForEntity);
```

Comment:
Look, my general feeling with this is that it is way more complex than it actually has to be. There is a single piece of logic (delete by query), but this is split into an two interfaces, two implementations, a "support class", uses lazy initialiation with locking, needs map/array look-ups etc. etc.

I would go for a much simpler design:

* Have `getLuceneQuery()` and `toString()` as instance methods on the query class (I don't mind splitting it up into interface/impl, so to allow for alternative impls down the road; you'd still need to figure a way though to configure which implementation to use, or is something for that already in place?)
* Have `fromString()` as a static method on the implementation class(es) (so you have your "everything-together" pattern)
* Invoke that static method during de-serialization if you see the right query key, otherwise raise an exception

That way you get rid of quite some code, making things much easier to understand and you still can evolve from there if actually needed.